packages/switch/src/switch.ts

```diff
@@ -15,31 +15,133 @@ import {
+    public defaultSelected?: boolean;
+
+    @property({ type: Boolean, reflect: true, attribute: 'checked' })
+    public isSelected?: boolean;
+
+    // react-stately will expect 'on' handlers for each state change
+    // this is where we apply changes to local state, and dispatch events
+    public onChange = (isSelected: boolean): void => {
+        const changeEvent = new CustomEvent('change', {
+            bubbles: true,
+            cancelable: true,
+        });
+        const apply = this.dispatchEvent(changeEvent);
+        if (apply) {
+            this.isSelected = isSelected;
+        }
+    };
+
+    @property({ reflect: true })
+    public value?: string;
+
+    @property({ reflect: true })
+    public name?: string;
+
+    // haunted specific: changes to lit-element, could be included in base class
+    private hauntedState = new State(() => this.requestUpdate(), this);
+
+    protected update(changedProperties: Map<string, boolean>): void {
+        this.hauntedState.run(() => super.update(changedProperties));
+    }
+
+    protected updated(changedProperties: PropertyValues): void {
+        super.updated(changedProperties);
+        this.hauntedState.runLayoutEffects();
+        defer(() => this.hauntedState.runEffects());
+    }
+
+    public disconnectedCallback(): void {
+        this.hauntedState.teardown();
+        super.disconnectedCallback();
+    }
+    // end haunted specific
+
+    // had to bring this up from checkbox base since we're not basing off that class here
+    @query('#input')
+    protected inputElement!: HTMLInputElement;
+
+    public get focusElement(): HTMLElement {
+        return this.inputElement;
+    }
+
     public static get styles(): CSSResultArray {
         /* istanbul ignore if */
         if (window.hasOwnProperty('ShadyDOM')) {
             // Override some styles if we are using the web component polyfill
-            return [...super.styles, switchStyles, legacyStyles];
+            return [switchStyles, legacyStyles];
         }
-        return [...super.styles, switchStyles];
+        return [switchStyles];
     }
 
     protected render(): TemplateResult {
+        // use the state from react-stately, this manages isSelected state
+        const state = useToggleState(this);
+        // use the hook from react-aria, this gives us event bindings and aria
+        // attributes to use in our DOM rendering
+        const { inputProps } = useSwitch(this, state, this);
+        // we have to hack the spread naming to be compatible with lit-html
+        fixSpreadNaming(inputProps);
+        // we also want to apply some of these attributes to our host (the aria ones)
+        for (const [key, value] of Object.entries(inputProps)) {
+            if (key.startsWith('aria-')) {
+                if (value) {
+                    this.setAttribute(key, value as any);
+                } else {
+                    this.removeAttribute(key);
+                }
+            }
+        }
```

Comment:
I think this part isn't actually working right, not seeing the aria attributes on the host correctly. Will need to dig into the exact logic needed here more if we decide to take this forward.

Reply:
One question I have is whether aria attributes should go on the host element or the `<input>` inside it? Seems like it should probably be the input, but then attributes like `aria-labelledby` wouldn't work because the id space is limited to the same shadow. Is there a way to make that work with shadow DOM?

Reply:
Applying these to the `<input>` is mostly the right answer. Currently, `CheckboxBase` is applying the `aria-` coverage we do have in this way and `Switch` is applying the `role` attribute similarly.

There are some cases where we've been cavalier or undecided about an `input`'s relationship to its `label`, but if we went this way it would be a good time to solidify this. Insights on how you saw this applying to slotted `input`s or elements with `elementInternals` would probably be a useful inclusion towards clarifying next steps here.

Reply:
Ah yes, sorry my misunderstanding then, probably don't need this loop at all to apply anything to the host then.

I think we should definitely explore how to apply to slotted inputs or elementInternals in this manner too. Probably some additional helpers around processing slotted content on slot change perhaps?